networking/src/p2p/peer/quota.rs

```diff
@@ -0,0 +1,199 @@
+                        default[i] = (tx, rx);
+                    }
+                }
+            }
+        }
+        default
+    };
+}
+
+fn decrease(q: &mut isize) {
+    *q = q.checked_sub(1).unwrap_or(*q)
+}
+
+pub(crate) struct ThrottleQuota {
+    quotas: [(isize, isize); THROTTLING_QUOTA_NUM],
+    quota_disabled: bool,
+    log: Logger,
+}
+
+impl ThrottleQuota {
+    pub(crate) fn new(log: Logger) -> Self {
+        Self {
+            quotas: THROTTLING_QUOTA_MAX.clone(),
+            quota_disabled: *THROTTLING_QUOTA_DISABLE,
+            log,
+        }
+    }
+
+    pub(crate) fn msg_index(msg: &PeerMessageResponse) -> usize {
+        debug_assert!(THROTTLING_QUOTA_NUM == 18);
+        match msg.message() {
+            PeerMessage::Disconnect => 0,
+            PeerMessage::Advertise(_) => 1,
+            PeerMessage::SwapRequest(_) => 2,
+            PeerMessage::SwapAck(_) => 3,
+            PeerMessage::Bootstrap => 4,
+            PeerMessage::GetCurrentBranch(_) => 5,
+            PeerMessage::CurrentBranch(_) => 6,
+            PeerMessage::Deactivate(_) => 7,
+            PeerMessage::GetCurrentHead(_) => 8,
+            PeerMessage::CurrentHead(_) => 9,
+            PeerMessage::GetBlockHeaders(_) => 10,
+            PeerMessage::BlockHeader(_) => 11,
+            PeerMessage::GetOperations(_) => 12,
+            PeerMessage::Operation(_) => 13,
+            PeerMessage::GetProtocols(_) => 14,
+            PeerMessage::Protocol(_) => 15,
+            PeerMessage::GetOperationsForBlocks(_) => 16,
+            PeerMessage::OperationsForBlocks(_) => 17,
+        }
+    }
+
+    fn index_to_str(index: usize) -> &'static str {
+        if index < THROTTLING_QUOTA_NUM {
+            THROTTLING_QUOTA_STRS[index]
+        } else {
+            "<invalid index>"
+        }
+    }
+
+    pub fn can_send(&mut self, msg: &PeerMessageResponse) -> bool {
+        let index = Self::msg_index(msg);
+        if THROTTLING_QUOTA_MAX[index].0 <= 0 {
+            return true;
+        }
+        decrease(&mut self.quotas[index].0);
+        if self.quota_disabled || self.quotas[index].0 >= 0 {
+            true
+        } else {
+            if self.quotas[index].0 == -1 {
+                warn!(self.log, "Cannot send message because its send quota is exceeded";
+                      "msg_kind" => Self::index_to_str(index));
+            }
+            false
+        }
+    }
+
+    pub fn can_receive(&mut self, msg: &PeerMessageResponse) -> bool {
+        let index = Self::msg_index(msg);
+        if THROTTLING_QUOTA_MAX[index].1 <= 0 {
+            return true;
+        }
+        decrease(&mut self.quotas[index].1);
+        if self.quota_disabled || self.quotas[index].1 >= 0 {
+            true
+        } else {
+            if self.quotas[index].1 == -1 {
+                warn!(self.log, "Cannot receive message because its receive quota is exceeded";
```

Comment:
@akoptelov 
I would also add it here peer_uri, that we know and analyze more

Reply:
We have that already, it is in the parent log:
```
Jul 19 16:54:25.219 WARN Cannot send message because its send quota is exceeded, msg_kind: Advertise, peer_uri: test-peer
Jul 19 16:54:30.170 WARN Tx quota is exceeded, amount: 11, quota: 1, msg_kind: Advertise, peer_uri: test-peer
```